File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        // Setting up our two pointers
        int left = 0, right = height.size() - 1;

        // Setting up to store the maximum calculated area
        int maxArea = 0;

        // Begin our two pointer approach
        while (left < right) {
            // The height is whatever column is smaller height wise
            int calcHeight = min(height[left], height[right]);

            // Getting the total distance between the two columns
            int width = right - left;

            // Calculating the current area with our calculations
            int area = calcHeight * width;

            // Updating our max area if it's greater
            maxArea = max(maxArea, area);

            // Advancing our pointers
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        // Returning the maximum area we calculated
        return maxArea;
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        stack = []
        maxArea = 0

        for i in range(len(heights)):
            while stack and heights[i] < heights[stack[-1]]:
                curIndex = stack.pop()

                height = heights[curIndex]
                right = i
                left = -1 if not stack else stack[-1]

                width = right - left - 1
                maxArea = max(maxArea, height * width)

            stack.append(i)

        while stack:
            curIndex = stack.pop()

            height = heights[curIndex]
            right = len(heights)
            left = -1 if not stack else stack[-1]

            width = right - left - 1
            maxArea = max(maxArea, height * width)

        return maxArea
